core/src/main/clojure/xtdb/expression.clj

```diff
@@ -1223,6 +1226,41 @@
    :->call-code (fn [[setting-name]]
                   `(current-setting (resolve-string ~setting-name)))})
 
+(defn sleep [millis]
+  (Thread/sleep millis))
+
+(defmethod codegen-call [:sleep :f64] [_expr]
+  (let [{convert :->call-code} (codegen-cast {:source-type :f64 :target-type [:duration :milli]})]
+    {:return-type :null
+     :->call-code (fn [lit]
+                    `(sleep ~(convert lit)))}))
+
+(defn string->duration [input]
```

Comment:
I'd:

* have this as `(defmethod codegen-cast [:utf8 :duration] ...)` so that anyone casting from string to duration gets to use it
* in the planner, you can then emit `(cast ~<expr> [:duration :millis])`
* `(defmethod codegen-call [:sleep :duration])` should then be the only one you need - looks like you can have both emitting to the same EE function